content/articles/fibonacci-memoization-time-complexity-in-javascript/index.md

```diff
@@ -0,0 +1,250 @@
+
+```js
+if(n in memo)return memo[n];
+// memo fetching logic
+```
+
+We used the original argument `n` as a key in our `memo` to return the value that correlates with the memo's key.
+
+> `(n in the memo)` condition is just some classic JavaScript syntax. So we are checking if some key is inside the JavaScript object.
+
+Let's store our entire result inside our memo; our key is `n`. We will complete our initial OPERATION and return what we have put in that memo. Therefore, we will not change any return values. 
+
+> Note that the access key should always be your argument.
+
+```js
+memo [n]= fib(n-1)-fib(n-2);
+return memo[n];
+```
+
+The next step is to ensure that all recursive function calls to access the same memo. We should pass a memo object to all of the recursive calls.
+
+```js
+memo[n]=fib(n-1,memo)+fib(n-2,memo);
+```
+
+We only receive a top-level `memo` object whenever we make a top-level call at a `fib` function. This is because we are not passing in a second argument as shown in the `return memo[n];`.
+
+Since we are not passing explicit second arguments in our recursive calls, they will receive the same `memo` object, and it will look like it's been passed by the reference.
+
+**Below is our complete memoized code:**
+
+```js
+const fib=(n, memo ={})=>{
+if(n in memo)return memo[n];
+```
+
+We added a new argument and a base case in the above code changes.
+
+```js
+if(n <=2) return 1;
+memo[n]=fib(n-1, memo) + fib(n-2, memo);
+return memo[n];
+```
+
+We also added a memo storing logic as `memo[n]=fib(n-1, memo) + fib(n-2, memo)`. Note that we did not substitute any of the functional logic. These changes, therefore, allow our program runtime to improve greatly.
+
+#### Tree implementation
+![memoized tree](/engineering-education/fibonacci-memoization-time-complexity-in-javascript/memoized.jpeg)
+
+In the above image, the red nodes return the memoized result. The algorithm is called once for each value from *0 to n*if they are ignored.
+
+A call tree for a figure more prominent than `fib(7)` will look just like the tree in the above image.
+
+#### Time complexity calculation 
+Let's take time complexity  of `n` to be`T(n)` , hence` T(n) = T(n-1) + T(n-2)`. 
+
+This is because `F(n-2)` is the one which is in the cache when we calculate `F(n - 1)`. Therefore, the function of `F(n-2)` is 1 (reading from cached equation), hence `T(n) = T(n-1) + 1 = T(n-2) + 2 = ... = T(n-n) + n.` 
+
+`T(0) is 1` will be regarded as `T(n) = O(n + 1) = O(n)`.
+
+### Conclusion
+The pattern of overlapping problems to reduce the time complexity of the Fibonacci series from *O(2n)* to *O(n)* is known as dynamic programming. The dynamic program will be any instance where we have some larger Fibonacci problem and can be divided into smaller instances of the same problem, thus achieving an overlapping structure.
```

Comment:
There are technical errors in this article. For example, reducing TC from O(2n) to O(n). I think that is incorrect. Can you correct it?

Reply:
Also, review and check the entire article for technical errors. If any, we will go ahead and close the article.